pwiz_tools/Skyline/Test/AntivirusExclusionTest.cs

```diff
@@ -37,7 +37,7 @@ namespace pwiz.SkylineTest
 
     public class AntivirusExclusionTest : AbstractUnitTest
     {
-        [TestMethod]
+        [TestMethod, NoParallelTesting] // No parallel testing as there's only one copy of the test file, so parallel agents may collide on access

```

Comment:
This makes a lot of sense, given that the Docker Containers will have no virus checking in C:\AlwaysUpCLT and therefore always pass this test. Instead, we want to ensure that it gets run at least once on the local disk drive.

Reply:
And this suggests that, yes, the docker container writes to a location other than the standard one, but not to a unique-per-client location:

!!! ConsoleMultiReplicateImportTest FAILED
Assert.Fail failed. No error reported but exit status was 2.
Output: 
Opening file...
File Imported_multiple2.sky opened.
160109_Mix1_calcurve_070 -> **c:\AlwaysUpCLT\TestResults\ImportAllCmdLineTest\160109_Mix1_calcurve_070.mzML**  Note: The file has already been imported. Ignoring...
REP01 -> c:\AlwaysUpCLT\TestResults\ImportAllCmdLineTest\REP01\CE_Vantage_15mTorr_0001_REP1_01.raw  Note: The file has already been imported. Ignoring...
Replicate 'REP01' already exists in the document, using 'REP012' instead.

Reply:
c:\AlwaysUpCLT is definitely a unique-to-container-instance location. Perhaps it's opening the document from a non-unique location (and another test has written to it there)?

Reply:
How commonly are you getting that ConsoleMultiReplicateImportTest failure? I'm running that test in an infinite loop, onscreen, all languages, without this PR, and I'm not seeing any failures.

Reply:
None of this is highly reproducible, other than TestAssayLibraryImport. I should dig in and see why that one is so reliable.

![image](https://user-images.githubusercontent.com/1200761/209864952-4b432e13-b23f-4537-ad1d-d5aca3dcc7fc.png)

Reply:
It is the most common intermittent failure for parallel testing. The failure is subtle, though. Note the very top of the message:

Assert.Fail failed. No error reported but exit status was 2.

This is saying that we got a failure exit status despite no true error being reported. Lots of pseudo-errors get reported, but they are of this form:

Warning: Failed importing the results file c:\AlwaysUpCLT\TestResults\ImportAllCmdLineTest\FullScan_folder\FullScan.RAW.  Ignoring...

Because the command-line specified to warn on error. The trick will be figuring out what is causing the exit code of 2 but not writing any error text to the output.

Reply:
Oh are those warnings like `Replicate 'REP01' already exists in the document, using 'REP012' instead.` expected to be there, so the real error is something not in the output?

Reply:
I don't understand docker behavior, but to the untrained eye it seems like these invocations all share a common value for the "results" argument in a way that other client-related arguments do not:

Launching docker_worker_0: docker run --name docker_worker_0 -it --rm -m 7914743398b -v c:\SkylineTestData:c:\downloads -v C:\Dev\master_clean:c:\pwiz chambm/always_up_runner "c:\pwiz\pwiz_tools\Skyline\bin\x64\Release\TestRunner.exe parallelmode=client showheader=0 results=c:\AlwaysUpCLT\TestResults log=c:\AlwaysUpCLT\TestRunner-docker_worker_0.log perftests=on teamcitytestdecoration=off buildcheck=0 runsmallmoleculeversions=on workerport=51659 workername=docker_worker_0" 
Launching docker_worker_1: docker run --name docker_worker_1 -it --rm -m 7914743398b -v c:\SkylineTestData:c:\downloads -v C:\Dev\master_clean:c:\pwiz chambm/always_up_runner "c:\pwiz\pwiz_tools\Skyline\bin\x64\Release\TestRunner.exe parallelmode=client showheader=0 results=c:\AlwaysUpCLT\TestResults log=c:\AlwaysUpCLT\TestRunner-docker_worker_1.log perftests=on teamcitytestdecoration=off buildcheck=0 runsmallmoleculeversions=on workerport=51659 workername=docker_worker_1" 
 [11:26] 0.0    FileTypeTest                                   (en)   0 failures, 2.93/9.15/70.8 MB, 2/486 handles, 2 sec.
Launching docker_worker_2: docker run --name docker_worker_2 -it --rm -m 7914743398b -v c:\SkylineTestData:c:\downloads -v C:\Dev\master_clean:c:\pwiz chambm/always_up_runner "c:\pwiz\pwiz_tools\Skyline\bin\x64\Release\TestRunner.exe parallelmode=client showheader=0 results=c:\AlwaysUpCLT\TestResults log=c:\AlwaysUpCLT\TestRunner-docker_worker_2.log perftests=on teamcitytestdecoration=off buildcheck=0 runsmallmoleculeversions=on workerport=51659 workername=docker_worker_2" 
Launching docker_worker_3: docker run --name docker_worker_3 -it --rm -m 7914743398b -v c:\SkylineTestData:c:\downloads -v C:\Dev\master_clean:c:\pwiz chambm/always_up_runner "c:\pwiz\pwiz_tools\Skyline\bin\x64\Release\TestRunner.exe parallelmode=client showheader=0 results=c:\AlwaysUpCLT\TestResults log=c:\AlwaysUpCLT\TestRunner-docker_worker_3.log perftests=on teamcitytestdecoration=off buildcheck=0 runsmallmoleculeversions=on workerport=51659 workername=docker_worker_3" 
 [11:26] 0.1    FileTypeTest                                   (zh)   0 failures, 2.94/9.15/70.7 MB, 3/491 handles, 2 sec.
Launching docker_worker_4: docker run --name docker_worker_4 -it --rm -m 7914743398b -v c:\SkylineTestData:c:\downloads -v C:\Dev\master_clean:c:\pwiz chambm/always_up_runner "c:\pwiz\pwiz_tools\Skyline\bin\x64\Release\TestRunner.exe parallelmode=client showheader=0 results=c:\AlwaysUpCLT\TestResults log=c:\AlwaysUpCLT\TestRunner-docker_worker_4.log perftests=on teamcitytestdecoration=off buildcheck=0 runsmallmoleculeversions=on workerport=51659 workername=docker_worker_4"

Reply:
Right. The reason the test is failing is that the command-line is returning an exit status of 2 without any output line that starts with "Error:", and I think the test is not expecting a line with "Error:" because it has specified to warn on import errors. But, all those seeming errors are expected to be there. No idea why the failure is intermittent.

Reply:
Brian the results= argument is to prevent it from using the default path which would be relative to the Skyline executable, which IS in a (subdirectory of a) shared location (`c:\dev\master_clean` on host, `c:\pwiz` in container).

Reply:
You have to scroll down a long way, but the important part of the test failure is the bottom stack trace:

   at Microsoft.VisualStudio.TestTools.UnitTesting.Assert.HandleFailure(String assertionName, String message)
   at Microsoft.VisualStudio.TestTools.UnitTesting.Assert.Fail(String message, Object[] parameters)
   at pwiz.SkylineTestUtil.AbstractUnitTestEx.ValidateRunExitStatus(Nullable`1 expectSuccess, Int32 exitStatus, Boolean errorReported, String consoleOutput) in C:\proj\pwiz\pwiz_tools\Skyline\TestUtil\AbstractUnitTestEx.cs:line 82
   at pwiz.SkylineTestUtil.AbstractUnitTestEx.RunCommand(Nullable`1 expectSuccess, String[] inputArgs) in C:\proj\pwiz\pwiz_tools\Skyline\TestUtil\AbstractUnitTestEx.cs:line 56
   at pwiz.SkylineTestData.CommandLineTest.ConsoleMultiReplicateImportTest() in C:\proj\pwiz\pwiz_tools\Skyline\TestData\CommandLineTest.cs:line 1498

And then CommandLineTest at line 1498:

            msg = RunCommand("--in=" + outPath2,
                             "--import-all=" + TestFilesDir.FullPath,
                             "--import-warn-on-failure",
                             "--save");

It is certainly confusing to read with the import failures reported as warnings with their own stack traces.

I can get it to fail in parallel testing pretty reliably, though. So, I will try to figure out what is going on.